### PR TITLE
1990N CH 5: Review/Submit

### PR DESCRIPTION
**Pull request by jbalboni**
Connects to [673](https://github.com/department-of-veterans-affairs/vets.gov-team/issues/673) and [2240](https://github.com/department-of-veterans-affairs/vets.gov-team/issues/2240)

This updates the confirmation page for 1990n. It also fixes a bug on the review page where required array fields prevent saving.

screen shot 2017-04-18 at 10 32 52 am

### CODE REVIEW

File: test/edu-benefits/5495/containers/ConfirmationPage.unit.spec.jsx [hunk 1]
```diff
@@ -0,0 +1,37 @@
+import React from 'react';
```
Comment (jbalboni) [inline test/edu-benefits/5495/containers/ConfirmationPage.unit.spec.jsx]: Forgot to check this file in in my last PR.

File: src/js/common/schemaform/helpers.js [hunk 2]
```diff
@@ -303,7 +303,15 @@ export function getNonArraySchema(schema) {
     }
 
     if (newProperties !== schema.properties) {
-      return _.set('properties', newProperties, schema);
+      let newSchema = _.set('properties', newProperties, schema);
+      if (newSchema.required) {
+        const newRequired = _.intersection(Object.keys(newSchema.properties), newSchema.required);
+        if (newRequired.length !== newSchema.required.length) {
+          newSchema = _.set('required', newRequired, newSchema);
```
Comment (cvalarida) [inline src/js/common/schemaform/helpers.js]: So...if these fields were required, but somehow they have no information in them, stop requiring them? Will this cause submission errors to the api?

File: src/js/edu-benefits/1990n/containers/ConfirmationPage.jsx [hunk 3]
```diff
@@ -52,15 +36,11 @@ class ConfirmationPage extends React.Component {
...
         </p>
         <div className="inset">
-          <h4>Education Benefit Claim <span className="additional">(Form 22-1990e)</span></h4>
+          <h4>Education Benefit Claim <span className="additional">(Form 22-1990n)</span></h4>
           <span>for {name.first} {name.middle} {name.last} {name.suffix}</span>
 
           <ul className="claim-list">
             <li>
-              <strong>Selected benefit</strong><br/>
-              {benefitsLabels[benefit]}
```
Comment (cvalarida) [inline src/js/edu-benefits/1990n/containers/ConfirmationPage.jsx]: Why is this removed?

File: src/js/edu-benefits/1990n/containers/ConfirmationPage.jsx [hunk 4]
```diff
@@ -52,15 +36,11 @@ class ConfirmationPage extends React.Component {
           <i>Please print this page for your records.</i>
         </p>
         <div className="inset">
-          <h4>Education Benefit Claim <span className="additional">(Form 22-1990e)</span></h4>
+          <h4>Education Benefit Claim <span className="additional">(Form 22-1990n)</span></h4>
```
Comment (cvalarida) [inline src/js/edu-benefits/1990n/containers/ConfirmationPage.jsx]: Heh, good catch!

File: test/edu-benefits/1990n/containers/ConfirmationPage.unit.spec.jsx [hunk 5]
```diff
@@ -0,0 +1,37 @@
...
+};
+
+describe('Edu 1990n <ConfirmationPage>', () => {
+  it('should render', () => {
+    const tree = SkinDeep.shallowRender(
+      <ConfirmationPage form={form}/>
+    );
+
+    expect(tree.subTree('.edu-page-title').text()).to.equal('Claim received');
+    expect(tree.everySubTree('span')[1].text().trim()).to.equal('for Jane Doe');
```
Comment (cvalarida) [inline test/edu-benefits/1990n/containers/ConfirmationPage.unit.spec.jsx]: I don't suppose there's a better way to find this element, is there?

**Review by cvalarida — COMMENTED** [review]
It looks good overall, but I'm concerned about removing the requirement on required fields with no data.

File: src/js/common/schemaform/helpers.js (hunk 2 above)
Comment (jbalboni) [inline src/js/common/schemaform/helpers.js]: This is due to how we deal with array fields on the review page. We pull them out of the page they're in and display them separately. Before, we were doing that but not removing the array fields from the required array, which caused the phantom validation error you saw. I won't cause any submission errors because we still validate everything before sending when you click submit. This change doesn't get saved anywhere, it's a temporary schema that's only used in the components on the review page.

It is another edge case, though, when you have an array field that requires at least one item. It's not really clear that the user has to enter an item if they come to the review page without one already. I opened an issue for us to think about simplifying how we handle array fields on the review page.

File: src/js/edu-benefits/1990n/containers/ConfirmationPage.jsx (hunk 3 above)
Comment (jbalboni) [inline src/js/edu-benefits/1990n/containers/ConfirmationPage.jsx]: You don't select an actual benefit on this form, just check a box saying pick the best one for me.

File: test/edu-benefits/1990n/containers/ConfirmationPage.unit.spec.jsx (hunk 5 above)
Comment (jbalboni) [inline test/edu-benefits/1990n/containers/ConfirmationPage.unit.spec.jsx]: We could probably put a class on it, but otherwise, I don't think so. Though I just copied this from an existing test and haven't take another look at it.

File: src/js/common/schemaform/helpers.js (hunk 2 above)
Comment (cvalarida) [inline src/js/common/schemaform/helpers.js]: So this will just let us update the page when reviewing, but will still fail validation when submitting?

Comment (jbalboni) [inline src/js/common/schemaform/helpers.js]: @cvalarida Correct.
